more detailled & corrected lua documentation of sound behavior

loffren174 opened the pull request:
- Goal of the PR
Improve the clarity of how the various parameters handed to minetest.sound_play() are affecting the end result.

This has come from my own need/curiosity to find answers, and realizing the documentation was very imcomplete, partially wrong*, and partially misleading. I spent the weekend running lots of tests (a LOT of parameters could potentially affect each other...) and collecting the pieces of code that are responsible for various behaviors, and created a lot of chat commands to directly test & verify the behavior in minetest 5.4.1, in minetestgame and devtest. This is the result, in the hope that it will help the next person reading the lua docs.

*Details on what is wrong and misleading will be commented in this pull request thread. 
 Most changes can be directly seen from the source code if you combine looking in 3-5 places, but most of it was a combination of reading the code and confirming the suspected behaviors by doing ingame verification tests, with a lot of result tables.

- Does it resolve any reported issue?
    Should resolve issue #11414 

- Does it relate to a topic on the roadmap?
   I assume accurate documentation is a soft goal anyway, even if not explicitly in the roadmap.

## To do
Formatting, choice of words. I'm quite certain of the results and doublechecked the cases where the old docs tells the opposite of what happens, but the style of the sentences may be improved. Also, if someone does the work of doublechecking my results on another system to rule out funky openAL behavior, or in case they don't believe me, that would be ideal. I will provide the mods (lists of chat commands) I used to test the parameters, to save them some time.

This PR is Ready for Review.

## How to test
I will list the most important places in the code that confirm this behavior, and upload a package of mods (a list of chat commands) which I used to note, organize and verify the behaviors, in a comment once I figure out how to load them onto github.

loffren174 commented:
This is an attempt to list the points where current documentation is a) misleading; b) partial, c) outright wrong.

**Point 1:** 
Missing documentation:
field 'fade' (SimpleSoundSpec value):
The 'fade' field is not mentioned in the current doc. However, it is fully functional when used by a SimpleSoundSpec in node definitions.
This adds the default value, the behavior at different values when it does not act as expected, and generally adds it to the list of parameters.

This can be tested with the attached mod:
'sss_pitch_fade_gain_test'
which provides several nodes with these parameters at values 0.1, 0.2, 0.5,1,2,5,10 as dig & punch sounds, for easy comparison.

Source: 
https://github.com/minetest/minetest/blob/master/src/sound.h#L29-L34

**Point 2:**
Missing documentation:
Maximum cap of 'gain' (SimpleSoundSpec value):
I did not find this behavior in code, I simply noted it when doing test runs on the 'gain' parameter. Any 'gain' above 1, when specified in the SimpleSoundSpec, is capped to 1. This has been found by comparing values of 0.1, 0.2, 0.5,1,2,5,10. All values above 1 have equal loudness on my system; all values between 0 and 1 decrease volume as expected.

This can be tested with the attached mod:
'sss_pitch_fade_gain_test'
which provides several nodes with these parameters at values 0.1, 0.2, 0.5,1,2,5,10 as dig & punch sounds, for easy comparison.

**Point 3:**
Missing documentation:
When using minetest.sound_play(), the values 'pitch','fade' of the SimpleSoundSpec are ignored.

This one really threw me for a loop, leading to a lot of time wasted on checking all corner cases. In the end, the code responsible for this behavior can be found at:
https://github.com/minetest/minetest/blob/master/src/server.cpp#L2115-L2123
What happens is that some values are first taken from 'params' (the parameter table), then values of 'spec' overwrite said values, AND THEN when sending the packet we again overwrite the received values with the ones found in 'params'. The end result is that any value 'pitch' or 'fade' provided in the SSS are ignored when using minetest.sound_play().

This can be read from the code mentioned above, or tested with the attached mods 'fade_parameter_test' and 'pitch_parameter_test'.

**Point 4:**
Missing documentation:
When using minetest.sound_play(), the value 'gain' from the SimpleSoundSpec is NOT discarded- it is multiplied with the gain provided in the 'params' table.

Source in code:
https://github.com/minetest/minetest/blob/master/src/server.cpp#L2119

This can also be tested with the attached mod 'gain_parameter_test'.

**Point 5:**
Partial/incomplete documentation:
The parameters in the 'params' table where not orderly listed, it was not clear on what values they may take (does to_player allow for several names? does exclude_player allow for several names? Two times no. But it wasn't mentioned anywhere.), and how they interact with the parameters in the SimpleSoundSpec, if both tables included said values. The interaction of parameters appearing in both table is tested in the earlier points already.

Some of the parameters in the 'params' table conflict, i.e. do not fit together, however the server still accepts them and acts in a particular manner. The behavior when several of these are provided was not explained anywhere (well, you can get it from combining two places of code.) Also, the parameters of the 'params' list were not clearly listed as a comprehensive list, the reader had to assume the parameters in the provided examples are all that exist.
The particular behavior/hierarchy in which the parameters override each other can be gathered from the combined reading of the following places:

Source in Code:
https://github.com/minetest/minetest/blob/master/src/server.cpp#L2053-L2100
https://github.com/minetest/minetest/blob/master/src/server.h#L94-L112

https://github.com/minetest/minetest/blob/master/src/script/common/c_content.cpp#L1030-L1097

These code pieces include all parameter behaviors EXCEPT the parameter 'loop', and to 'max_hear_distance' to the extent that the 'loop' parameter affects it.
I could not find anything directly in the code pertaining to the loop behavior, but I stopped once I tracked it to openAL getting passed a 'loop' variable, and continued by directly testing in the client, and instead ran direct tests in the client.

The behaviors of the parameters 'loop' and 'max_hear_distance', and their surprising combination, can be tested with the attached mods 'max_hear_distance_parameter_test' and 'loop_parameter_test'.

This leads to Point 6:

**Point 6:**
Wrong documentation:

In the current lua doc:
https://github.com/minetest/minetest/blob/master/doc/lua_api.txt#L889-L890

"Looped sounds must either be connected to an object or played locationless to
one player using `to_player = name`

This is wrong according to my tests. And because I did not think the docs could be wrong, I ran through them four times. The current situation is that all four modes - to_player,pos,object,global - support a 'loop=true' parameter and the sound actually loops.

This can be tested with the attached mod 'loop_parameter_test'.

**Point 7:**
Partial/incomplete documentation:
In the current lua doc:
https://github.com/minetest/minetest/blob/master/doc/lua_api.txt#L892-L893

"
A positional sound will only be heard by players that are within
`max_hear_distance` of the sound position, at the start of the sound."

This is true. However, this comment and the one above, are the only parts mentioning the looped sounds behavior. Since the behavior is very peculiar, this information is fragmented at best, and does not include important behaviors, such as:

The sound continues to loop once the initial sound is heard, no matter how far the player goes away, disregarding the max_hear_distance after the initial sound.

When attached to an object, the looped sound follows the object's position with the looped sound's 'center' as long as it exists, but keeps playing on its last position after the object has been removed.


**Point 8:**
Partial/incomplete documentation:

In the current lua doc:
https://github.com/minetest/minetest/blob/master/doc/lua_api.txt#L895-L896

"`exclude_player = name` can be applied to locationless, positional and object-
bound sounds to exclude a single player from hearing them."

This mentions three of the four modes of sound - locationless (global), positional, and object-bound. It doesn't mention 'to_player', so the assumption the reader gets is that to_player overrides exclude_player. This is correct- but it is much clearer to instead detail the entire hierarchy of parameters, and which override which, than only mentioning the relation of two parameters partially, implying the 4th mode of sound for them.

As such, all of the last three comments can be removed in favor of the new explanation. The second comment, regarding positional sound and max_hear_distance, is valid and 100% correct and thus has been kept, even if it may be redundant now since this explanation is included in the two new paragraphs before it.

----------------------------------------------------------------------

This should adress all changes in this PR.

loffren174 commented:
I uploaded the mods (just a list of chat commands, as said before) which I used to determine the behavior, so they may save someone time when doublechecking results, although not that much. The intended way of using the folders is to load them as minetest mods, and try out the chat commands while simultaneously reading the source .lua file of the chat commands.

The zip file can be found here:
[https://github.com/loffren174/minetest_sound_tests_mods_and_results/blob/main/tests_and_results.zip](url)

Wuzzy2 commented:
This is amazing. Oh my! I didn't realize the sound API is so messy. Assuming your new documentation is correct, this looks like several design flaws, I find it all pretty confusing to be honest with all the awkward special cases that don't seem to serve any purpose. Especially the overriding rules don't seem to make any sense. But anyway, you're just documenting, it's obviously not your fault and it's obviously better to have that awkwardness documented.

One complaint about your documentation: You seem to mix tabs and spaces at some places, please fix this.

Have you noticed in your tests that the range that positional sounds become "more silent" the further away you go is fixed, and doesn't scale properly with`max_hear_distance`? By default, you reach zero volume after `max_hear_distance`, and it is smooth. But if `max_hear_distance` is halved, you will go from 50% volume to 0% as you cross the `max_hear_distance`, there's no smoothness at all.

Wuzzy2 commented:
One thing that still confuses me is why `gain`, `pitch` and `fade` are duplicated in SimpleSoundSpec and Sound Parameters. I don't see the point, especially what it means for development. Maybe it would help to provide some 'best practices' or something like that.

> While passing a `pitch` and `fade` value in Sound Parameters is optional, the 
SimpleSoundSpec's `pitch` and `fade` are ignored by minetest.sound_play(), so if said 
values are not the default '1','0' then it is required to pass the same values 
again in the Sound Parameters table, to have identical results to using the SimpleSoundSpec in 
a node definition.

That is just so messed up, lol. Not your fault, of course.

I think about `fade` you should say what it means. I believe the number is the time in seconds, right?

> The table above is ordered 
> in descending priority of arguments overriding each other: 
> `to_player`,`exclude_player`,`pos`,`object`,`max_hear_distance`.

Wait, what? So you're saying, that if any of `to_player`,`exclude_player`,`pos`,`object` is provided, then `max_hear_distance` is ignored??? No, that's not correct. `max_hear_distance` definitely works if `pos` is provided.

> However, players who heard the initial sound will continue to hear the looped sound at a minimum volume even if they leave beyond `max_hear_distance`

This is 100% a bug and should be stated as such ("quirk" is too soft a word). I believe we have an issue for that somewhere …

loffren174 commented:
> Wait, what? So you're saying, that if any of to_player,exclude_player,pos,object is provided, then max_hear_distance is ignored??? No, that's not correct. max_hear_distance definitely works if pos is provided.

No, sorry. I tried to cut down the explanation to be as short as possible, so I potentially included some misunderstandings again. The only parameter of those mentioned which _overrides_ max_hear_distance is to_player. I will put that into an extra note, below the one about exclude_player only overriding for the specified player.

> I think about fade you should say what it means. I believe the number is the time in seconds, right?

I don't *know* the precise meaning. I would have to re-test this with some longer sound when I have time, but from my initial tests, I don't think it's as straightforward as this, because a value of '10' is still quieter even on the noteblock sound which is way shorter than 10 seconds. This is same as the 'gain' cap - I'd welcome someone who knows how to dig in the code, potentially the audio library we use, to confirm what happens with those parameters there. But judging from my own listening tests, I don't think it's as straightforward as just cutting off the length in seconds, as even using the values '5' or '10' results in a noticably 'quieter' sound at the relatively short sound of mesecons_noteblock_a.ogg than using the value '0' (disable). Maybe it's in tenths of seconds, or the fade has some early period of effect ahead of the specified 'time', or it depends on something else related to volume or- what I suspect - it's a combination of one or two of these things.

Wuzzy2 commented:
> The only parameter overriding `max_distance` is `to_player`.

Lol, it's not `max_distance` ... I feel like the entire section describing overriding should be rewritten completely. It turns out that unfortunately, it is NOT a simple hierarchy or order of preference as you portray it. I sugget to portray this overriding stuff in some other way, with no contradictions.

It is OK to simplify, but only as long you don't lie. :D

> Frankly, I assume a lot of the things noted here to be in a grey undefined zone

I think that's OK. Making the documentation correct is a good thing either way. As I said, it's not your fault. Especially for this weird API a good documentation is very important.

loffren174 commented:
I ran some more tests according to the questions you asked, added some notes, and added the changes at the end of the second post in this thread. Please note the changes, as I corrected an initial assumption that turned out more precise than the current doc, but still wrong (regarding volume of looped positional sounds.)

> I feel like the entire section describing overriding should be rewritten completely. It turns out that unfortunately, it is NOT a simple hierarchy or order of preference as you portray it. I sugget to portray this overriding stuff in some other way, with no contradictions.

Originally, I had this awful list of notes of branching IF-THEN conditions, which I then thought how to portray more compact... I agree that the depiction now is not intuitive, but I didn't want to add so many more paragraphs essentially paraphrasing the code... I am very much open for suggestions on this?

The code in question (as mentioned before already):
Source in Code:
https://github.com/minetest/minetest/blob/master/src/server.cpp#L2053-L2100
https://github.com/minetest/minetest/blob/master/src/server.h#L94-L112
https://github.com/minetest/minetest/blob/master/src/script/common/c_content.cpp#L1030-L1097

My summary notes:

_So, the decision chain is as follows, in this order:
if a to_player parameter is given, all pos, max_distance,object fields are ignored and the sound is played to the player. The sound even plays if the same player is passed in the exclude_player parameter.
if an exclude_player parameter is given, they will not hear it in any case except when they are also passed as to_player.
if no to_player param is given, and a pos param is given, the sound will be played at pos with the max_hear_distance given, or the default max_hear_distance if none was given. (32nodes)
if no to_player and no pos parameter was given and an object parameter was given, the sound will be played at the object's position, with the same behavior regarding max_hear_distance. If said position can not be retrieved (because the object does not exist anymore), the sound is not played.
if no to_player, no pos, and no object parameter is given, the sound plays globally to all players (except the player parameter passed in exclude_player)._

appgurueu commented on doc/lua_api.txt:
```diff
@@ -845,6 +845,131 @@ from the available ones of the following files:
+This table has the following parameters, all of which are optional, same as the 
+table itself (when omitted, default values are used for the values that have them):
+
+{
```
Your formatting here is severely broken - you only semi-converted the Lua code snippet to Markdown...

Wuzzy2 commented:
Formatting is indeed very broken. Here you can test your Markdown: https://daringfireball.net/projects/markdown/dingus

Also make sure all variable/function names are in backticks (unless already in a code section) to get the "code style".

> The `fade`, `gain` parameters correspond to the 'step', 'gain' parameters of 
> `minetest.sound_fade`

I don't understand this section. Correspond in what way?

> The `loop` parameter has several quirks: It does not override `max_hear_distance`, which 
> decides whether the initial sound is heard.

This actually sounds more like a bug than like a quirk.

appgurueu commented on doc/lua_api.txt:
```diff
@@ -845,6 +845,131 @@ from the available ones of the following files:
+
+The `loop` parameter has several quirks: It does not override `max_hear_distance`, which 
+decides whether the initial sound is heard. However, players who heard the initial sound 
+will continue to hear the looped sound at a low volume according to their distance from it,
```
Couldn't this entire section be abbreviated to "even if the distance changes, this does not affect the volume of the sound"?

loffren174 commented:
I am closing this in favour of PR https://github.com/minetest/minetest/pull/12764/

reasons to close this:
1. changes to sound code since creating this PR would require comprehensive re-testing
2. keeping odd behaviours undocumented makes it easier to consecutively fix them without making it a breaking change
3. improving behaviour synchronously with updating the documentation is preferable to updating documentation alone
4. sparse available dev attention should instead be redirected at such superior approaches, as in PR 12764

I apologize for the discussion and the PR fizzing out up to now, likely due to me not approaching things with the right attitude and in a timely manner. Thanks to everyone involved!

Desour commented:
>sparse available dev attention

Sorry for not reviewing your PR.
My reason for this was that I didn't want you to do unnecessary work. (After all, your approach of finding all this stuff out just by testing is very difficult and error-prone.)
Anyway, thanks for your efforts! If you find any unclear or missing points in the documentation in #12764, please complain.
